src/request/reducers/RequestDetailRequestReducer.ts

```diff
@@ -19,42 +26,156 @@ const defaultState: IRequestDetailRequestState = {
+        messageStack.push(middleware);
+    });
+
+    return messageStack[0].middleware;
 }
 
-export function requestReducer(state: IRequestDetailRequestState = defaultState, action: Action) {
+function toMap<T, TResult>(values: T[], keySelector: (value: T) => string, valueSelector: (value: T) => TResult): { [key: string]: TResult } {
+
+    return _.reduce(
+        values, 
+        (result: { [key: string]: TResult }, value: T) => {
+            result[keySelector(value)] = valueSelector(value);
+
+            return result;
+        }, 
+        <{ [key: string]: TResult }>{});
+}
+
+function createMiddlewareState(messages: ICorrelatedMiddlewareMessages): IRequestDetailRequestMiddlewareState {
+    // NOTE: We ignore Express Router header modifications as they're likely actually modifications made by route middleware, not the Router itself.
+
+    return {
+        headers: messages.endMessage && messages.startMessage.payload.name !== 'router' ? toMap(messages.endMessage.payload.headers, header => header.name, header => header.value) : {},
+        middleware: messages.middleware.map(middlewareMessages => createMiddlewareState(middlewareMessages)),
+        name: messages.startMessage.payload.displayName || messages.startMessage.payload.name,
+        packageName: messages.startMessage.payload.packageName
+    };
+}
+
+function updateMiddlewareState(request): IRequestDetailRequestMiddlewareState[] {
+    return correlateMiddlewareMessages(getMessages(request, MiddlewareStartType), getMessages(request, MiddlewareEndType)).map(messages => createMiddlewareState(messages));
+}
+
+export function middlewareReducer(state: IRequestDetailRequestMiddlewareState[] = defaultMiddlewareState, action: Action): IRequestDetailRequestMiddlewareState[]  {
     switch (action.type) {
         case requestDetailUpdateAction.type:
-            return updateRequestState(state, requestDetailUpdateAction.unwrap(action));
+            return updateMiddlewareState(requestDetailUpdateAction.unwrap(action));
     }
 
     return state;
 }
+
+function createRequestReducer<T>(defaultState: T, reducer: (state: T, requestPayload: IWebRequestPayload, responsePayload: IWebResponsePayload) => T): (state: T, action: Action) => T {
+    return (state: T = defaultState, action: Action) => {
+        switch (action.type) {
+            case requestDetailUpdateAction.type: {
+                const request = requestDetailUpdateAction.unwrap(action);
+                const requestMessage = getMessageWithPayload<IWebRequestPayload>(request, WebRequestType);
+                const responseMessage = getMessageWithPayload<IWebResponsePayload>(request, WebResponseType);
+
+                return reducer(
+                    state, 
+                    requestMessage ? requestMessage.payload : undefined,
+                    responseMessage ? responseMessage.payload : undefined);
+            }
+        }
+
+        return state;
+    }
+}
```

Comment:
To accomodate the addition of a middleware reducer (in particular, to allow testing of the middleware reducer apart from the other request-related reducing), I needed to split apart that "reduction" into each constituent parts (i.e. url, request, and response properties) so they could then be combined with middleware.  Each of those reductions had the same form (i.e. responded to the same actions) so I created a factory (`createRequestReducer`) that generates the boilerplate for handling those actions.